File: frontend/src/types/calculation-step.ts
/**
 * CalculationStep domain types for audit trail of reconciliation calculations.
 *
 * These Zod schemas match exactly with backend/app/models/calculation_step.py.
 * Each step captures inputs, operation, output, and optional notes for audit purposes.
 */

import { z } from 'zod'
import { formatMoney } from '@/lib/money'

/**
 * Schema for output values - can be either a decimal string or a complex object.
 */
export const OutputValueSchema = z.union([
  z.string().regex(/^-?\d+(\.\d+)?$/, 'Must be a valid decimal number'),
  z.record(z.string(), z.unknown()),
])

/**
 * Schema for input values - flexible dict structure.
 * Must have at least one key.
 */
export const InputValuesSchema = z
  .record(z.string(), z.unknown())
  .refine((obj) => Object.keys(obj).length > 0, {
    message: 'input_values cannot be empty',
  })

/**
 * Full CalculationStep model for audit trail.
 *
 * Represents a single step in a reconciliation calculation,
 * capturing what inputs were used, what operation was performed,
 * and what output was produced.
 */
export const CalculationStepSchema = z.object({
  step_order: z.number().int().min(1, 'step_order must be at least 1'),
  step_name: z
    .string()
    .min(1, 'step_name is required')
    .max(100, 'step_name must be at most 100 characters'),
  input_values: InputValuesSchema,
  /** Per-input unit tags. Keys absent here default to 'currency' on display. */
  input_units: z.record(z.string(), z.string()).optional(),
  operation: z
    .string()
    .min(1, 'operation is required')
    .max(500, 'operation must be at most 500 characters'),
  output_value: OutputValueSchema,
  /** Unit tag for output_value. Defaults to 'currency' when absent. */
  output_unit: z.string().optional(),
  note: z
    .string()
    .max(500, 'note must be at most 500 characters')
    .nullable()
    .optional(),
})

export type CalculationStep = z.infer<typeof CalculationStepSchema>

/**
 * DTO for creating a calculation step.
 * Same as CalculationStep since all fields except note are required.
 */
export const CalculationStepCreateSchema = CalculationStepSchema

export type CalculationStepCreate = z.infer<typeof CalculationStepCreateSchema>

/**
 * Factory function to create a calculation step object.
 *
 * @param step_order - Order of this step (1-indexed)
 * @param step_name - Human-readable name for the step
 * @param input_values - Dict of inputs used in calculation
 * @param operation - Formula or description of operation
 * @param output_value - Result of the calculation
 * @param note - Optional explanation or warning
 * @returns A validated CalculationStep object
 */
export const createCalculationStep = (
  step_order: number,
  step_name: string,
  input_values: Record<string, unknown>,
  operation: string,
  output_value: string | Record<string, unknown>,
  note?: string | null
): CalculationStep => {
  return CalculationStepSchema.parse({
    step_order,
    step_name,
    input_values,
    operation,
    output_value,
    note: note ?? null,
  })
}

/**
 * Format a calculation step as a human-readable summary.
 *
 * @param step - The calculation step to format
 * @returns A formatted string summary of the step
 */
export const formatStepSummary = (step: CalculationStep): string => {
  let outputStr: string
  if (typeof step.output_value === 'string') {
    const num = parseFloat(step.output_value)
    if (!isNaN(num)) {
      outputStr = formatMoney(num)
    } else {
      outputStr = step.output_value
    }
  } else {
    outputStr = JSON.stringify(step.output_value)
  }

  let summary = `Step ${step.step_order}: ${step.step_name} = ${outputStr}`
  if (step.note) {
    summary += ` (${step.note})`
  }
  return summary
}

/**
 * Validate that a list of calculation steps has correct sequential ordering.
 *
 * @param steps - List of calculation steps to validate
 * @returns True if steps are correctly ordered (1, 2, 3, ...)
 * @throws Error if steps are not in sequential order starting from 1
 */
export const validateStepSequence = (steps: CalculationStep[]): boolean => {
  if (steps.length === 0) {
    return true
  }

  let expectedOrder = 1
  for (const step of steps) {
    if (step.step_order !== expectedOrder) {
      throw new Error(
        `Expected step_order ${expectedOrder}, got ${step.step_order}`
      )
    }
    expectedOrder++
  }

  return true
}

/**
 * Get a short description of what a step does based on its name.
 *
 * @param stepName - The step name to describe
 * @returns A short description or the original name if no match
 */
export const getStepDescription = (stepName: string): string => {
  const descriptions: Record<string, string> = {
    'Calculate Actual Occupancy': 'Determines current building occupancy rate',
    'Calculate Gross-Up Factor': 'Adjusts expenses for occupancy level',
    'Apply Gross-Up': 'Applies gross-up factor to variable expenses',
    'Calculate Base Year Amount': 'Retrieves or calculates base year expenses',
    'Calculate Tenant Share': "Computes tenant's pro-rata share of expenses",
    'Apply Cap': 'Limits increase based on cap provisions',
    'Calculate Admin Fee': 'Computes administrative fee amount',
    'Calculate Total Recovery': 'Sums all recoverable amounts',
  }
  return descriptions[stepName] || stepName
}

/**
 * Check if a step has a warning note.
 *
 * @param step - The calculation step to check
 * @returns True if the step has a note indicating a warning
 */
export const hasWarning = (step: CalculationStep): boolean => {
  if (!step.note) {
    return false
  }
  const warningKeywords = ['warning', 'exceeded', 'capped', 'limit', 'adjusted']
  return warningKeywords.some((keyword) =>
    step.note!.toLowerCase().includes(keyword)
  )
}

File: backend/app/models/calculation_step.py
"""
CalculationStep domain type for audit trail of reconciliation calculations.

Each CalculationStep represents a single step in the reconciliation calculation
process, capturing inputs, the operation performed, and the resulting output.
This enables complete audit trail and debugging of financial calculations.
"""

from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.services.formatting import format_usd


class CalculationStep(BaseModel):
    """
    A single step in a reconciliation calculation.

    Used for audit trail and debugging. Each step captures:
    - What inputs were used
    - What operation was performed
    - What output was produced
    - Any notes or warnings

    Example:
        CalculationStep(
            step_order=1,
            step_name="Calculate Actual Occupancy",
            input_values={"occupied_sqft": 45000, "total_sqft": 50000},
            operation="occupied_sqft / total_sqft",
            output_value=Decimal("0.90"),
            note=None
        )
    """

    step_order: int = Field(
        ...,
        ge=1,
        description="Order of this step in the calculation sequence (1-indexed)",
    )
    step_name: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Human-readable name for this calculation step",
    )
    input_values: dict[str, Any] = Field(
        ...,
        description="Input values used in this step",
    )
    input_units: dict[str, str] = Field(
        default_factory=dict,
        description=(
            "Unit tag per input key (currency/ratio/area/count/date/text). "
            "Keys absent here default to currency on display."
        ),
    )
    operation: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Formula or description of the operation performed",
    )
    output_value: Decimal | dict[str, Any] = Field(
        ...,
        description="Result of step (Decimal or dict for complex values)",
    )
    output_unit: str = Field(
        default="currency",
        description="Unit tag for output_value (defaults to currency).",
    )
    note: str | None = Field(
        None,
        max_length=500,
        description="Explanation, warning, or clarification about this step",
    )

    @field_validator("input_values")
    @classmethod
    def validate_input_values_not_empty(cls, v: dict[str, Any]) -> dict[str, Any]:
        """Ensure input_values is not empty."""
        if not v:
            raise ValueError("input_values cannot be empty")
        return v


class CalculationStepCreate(BaseModel):
    """
    DTO for creating a calculation step.

    All fields except note are required.
    """

    step_order: int = Field(..., ge=1)
    step_name: str = Field(..., min_length=1, max_length=100)
    input_values: dict[str, Any] = Field(...)
    input_units: dict[str, str] = Field(default_factory=dict)
    operation: str = Field(..., min_length=1, max_length=500)
    output_value: Decimal | dict[str, Any] = Field(...)
    output_unit: str = Field(default="currency")
    note: str | None = Field(None, max_length=500)

    @field_validator("input_values")
    @classmethod
    def validate_input_values_not_empty(cls, v: dict[str, Any]) -> dict[str, Any]:
        """Ensure input_values is not empty."""
        if not v:
            raise ValueError("input_values cannot be empty")
        return v


def create_calculation_step(
    step_order: int,
    step_name: str,
    input_values: dict[str, Any],
    operation: str,
    output_value: Decimal | dict[str, Any],
    note: str | None = None,
) -> CalculationStep:
    """
    Factory function to create a CalculationStep.

    Args:
        step_order: Order of this step (1-indexed)
        step_name: Human-readable name for the step
        input_values: Dict of inputs used in calculation
        operation: Formula or description of operation
        output_value: Result of the calculation
        note: Optional explanation or warning

    Returns:
        A validated CalculationStep instance
    """
    return CalculationStep(
        step_order=step_order,
        step_name=step_name,
        input_values=input_values,
        operation=operation,
        output_value=output_value,
        note=note,
    )


def format_step_summary(step: CalculationStep) -> str:
    """
    Format a calculation step as a human-readable summary.

    Args:
        step: The calculation step to format

    Returns:
        A formatted string summary of the step
    """
    output_str = (
        format_usd(step.output_value)
        if isinstance(step.output_value, Decimal)
        else str(step.output_value)
    )
    summary = f"Step {step.step_order}: {step.step_name} = {output_str}"
    if step.note:
        summary += f" ({step.note})"
    return summary


def validate_step_sequence(steps: list[CalculationStep]) -> bool:
    """
    Validate that a list of calculation steps has correct sequential ordering.

    Args:
        steps: List of calculation steps to validate

    Returns:
        True if steps are correctly ordered (1, 2, 3, ...)

    Raises:
        ValueError: If steps are not in sequential order starting from 1
    """
    if not steps:
        return True

    expected_order = 1
    for step in steps:
        if step.step_order != expected_order:
            raise ValueError(
                f"Expected step_order {expected_order}, got {step.step_order}"
            )
        expected_order += 1

    return True
